### ISSUE TITLE
Support multi-format `next.config.{js,cjs,mjs,ts}`

### ISSUE THREAD

**Issue by 1aron** [vercel/next.js#44324]
### Describe the feature you'd like to request

- next.config.js
- next.config.mjs
- next.config.cjs
- next.config.ts

## Related

- https://github.com/vercel/next.js/discussions/32239
- https://github.com/vercel/next.js/pull/29935
- https://github.com/vercel/next.js/issues/34448
- https://github.com/vercel/next.js/pull/34451
- https://github.com/vercel/next.js/pull/33637
- https://github.com/vercel/next.js/pull/22153

### Describe the solution you'd like

As far as is generally known:
- Cannot import ESM modules in CJS
- Cannot use ESM dynamic import in CJS
- Cannot use CJS dynamic require in ESM
- Cannot import .ts config in .js .cjs .mjs
- Cannot mix .js .cjs .mjs packages

Recently, I encountered the same problem when implementing the [compiler of Master CSS](https://github.com/master-co/css/tree/beta/packages/compiler), so I created a plugin — [cross-import](https://github.com/1aron/utils/tree/main/packages/cross-import) that **allows you to import ESM, Typescript modules in the CJS environment at runtime**.

The principle is to **compile, bundle and create virtual JavaScript modules at runtime through ESBuild**, which is fast and formal because it is not hardcoded through `fs`.

Also, [cross-import](https://github.com/1aron/utils/tree/main/packages/cross-import) is not asynchronous like ESM's Dynamic Import, it runs synchronously.

So now Next Team supports multi-format JavaScript modules in a CJS environment almost immediately:

```bash
npm i cross-import
```

```ts
import crossImport from 'cross-import'

const config = crossImport('next.config.{js,cjs,mjs,ts}')
```
By default `cross-import` is executed in CWD.

### Describe alternatives you've considered

I also created a package — [techor](https://github.com/1aron/techor) to handle this kind of tech authoring.

```ts
const defaultOptions = {}

class NextJS extends Techor<Options, Config> {
    constructor (
        options: Options
    ) {
        super(defaultOptions, options)
    }
    ...
}
```
I think this is quite possibly the only solution for different environment JavaScript modules!

**Comment by Fredkiss3** [comment]
Right now, you can use the formats : `next.config.js` and `next.config.mjs` for the other I didn’t test but I suppose you can use them ?
I suspect that the `.ts` one is not possible because it is directly executed as a node argument without typescript ( something like : node `next.config.js` ), so it is not compiled before execution.

**Comment by 1aron** [comment]
> Right now, you can use the formats : `next.config.js` and `next.config.mjs` for the other I didn’t test but I suppose you can use them ? I suspect that the `.ts` one is not possible because it is directly executed as a node argument without typescript ( something like : node `next.config.js` ), so it is not compiled before execution.

The code is very short, you can understand how `ts` is imported as a module from this context: https://github.com/1aron/utils/blob/main/packages/cross-import/src/index.ts#L18-L26
It's possible.

**Comment by ondrej-langr** [comment]
You can even import things with `ts-node` like this:
```mjs
import tsNode from 'ts-node';
// This registers new ts compiler onto node itself which results in valid import of `.ts` files
tsNode.register();
const someFileModule = await import("some-file.ts")
```

**Comment by GabenGar** [comment]
Yeah but then you'll have an issue of whatever real-time config transpiler having a different typscript config than nextjs.
